ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -843,19 +844,64 @@ def _get_base_url(self, **kwargs: Dict[str, Any]) -> Optional[str]:
 
     def handle_prompt_template_invoke(self, instance, result, args: List[Any], kwargs: Dict[str, Any]):
         """On prompt template invoke, store the template on the result so its available to consuming .invoke()."""
-        template, variables = None, None
+        chat_template, template, variables = None, None, None
         if hasattr(instance, "template") and isinstance(instance.template, str):
             template = instance.template
+
+        if (
+            isinstance(getattr(instance, "messages", None), list)
+            and result is not None
+            and isinstance(getattr(result, "messages", None), list)
+        ):
+            messages = []
+            if len(instance.messages) != len(result.messages):
+                # langchain allows for message placeholder templates, but we don't support them yet
+                # these are templates where a template variable takes as its value several complete messages
+                log.debug(
+                    "Instance messages and result messages have different lengths; message placeholder not supported"
+                )
+                return
+
+            for m, r in zip(instance.messages, result.messages):
+                # message templates do not have a role: the role is derived from the instance class
+                # instead, we use the resulting message
+                role = getattr(r, "role", None)
+                if not role and hasattr(r, "type"):
+                    role = LANGCHAIN_ROLE_MAPPING.get(r.type)
+                if not role:
+                    role = "unknown"
+
+                if hasattr(m, "content"):
+                    # a message in the template that is not a template itself
+                    messages.append({"role": role, "content": m.content})
+                elif hasattr(m, "prompt") and hasattr(m.prompt, "template"):
+                    # a template message
+                    messages.append({"role": role, "content": str(m.prompt.template or "")})
+                else:
+                    messages = []
+                    log.debug("Failed to parse template messages")
+                    break
+            chat_template = messages if messages else None
+
         variables = get_argument_value(args, kwargs, 0, "input", optional=True)
+        if (
+            isinstance(variables, str)
+            and hasattr(instance, "input_variables")
+            and isinstance(instance.input_variables, list)
+            and len(instance.input_variables) == 1
+        ):
+            # variables should be passed in as a dict, but a string is allowed if there is only one input variable
+            variables = {instance.input_variables[0]: variables}
 
-        if not template or not variables or not isinstance(variables, dict):
+        if (not template and not chat_template) or not variables or not isinstance(variables, dict):
```

Comment:
Correct me if I'm misunderstanding here, but we are basically either populating the `template` var in the first case, or the `chat_template` var if that's not true? And we should never be setting both right? If that's the case should we also return if both are set (currently we're only returning if neither are set?)